<a href="https://colab.research.google.com/github/Noks06/Prediction-of-Product-Sales/blob/main/Sales_Predictions_Project_Core_Part(5)_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sales Predictions Project 1 (Core): Part 5
 By Nokuthula Mchunu

## Loading and Explore Data

In [191]:
from numpy.core.overrides import set_array_function_like_doc
#importing pandas and numpy
import pandas as pd
import numpy as np
# Importing sklearn liberaries and tools
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import make_pipeline
from sklearn import set_config
set_config(transform_output='pandas')

# Reading csv file to file variable fpath
fpath  = '/content/drive/MyDrive/CodingDojo/02-MachineLearning/Week05/Data/sales_predictions_2023.csv'
# load the dataframe
df = pd.read_csv(fpath)
# explore top 5 records
df.head(5)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [192]:
#Checking dataframe info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


* The info above tells us that there are 8523 records, 12 columns.  
* 4 float64, 1 int64 and 7 object
* Item_Weight and Outlet_size have missing values


## Perform Basic EDA


### Check for Duplicate

In [193]:
# check if there are duplicate

df.duplicated().sum()

0

there are no duplicates

### Checking inconsistencies

In [194]:
# Checking inconsistancy in Numeric columns
df.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      7060.00          8523.00   8523.00                    8523.00   
mean         12.86             0.07    140.99                    1997.83   
std           4.64             0.05     62.28                       8.37   
min           4.56             0.00     31.29                    1985.00   
25%           8.77             0.03     93.83                    1987.00   
50%          12.60             0.05    143.01                    1999.00   
75%          16.85             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96

Numeric values seems to be fine, there are no odd numbers or out of range but there are missing values on Item_weight

In [195]:
#Checking inconsitancy in Nominal Columns
cat_col = df.select_dtypes('object').columns
cat_col


Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [196]:
# Checking the value counts of all categorical  columns
for col in cat_col:
  print(df[col].value_counts())
  print('')

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

Medium   

In [197]:
# Correcting the inconsistancy by using a replace method
df['Item_Fat_Content']= df['Item_Fat_Content'].replace({'LF':'Low Fat','low fat':'Low Fat','reg':'Regular'})
# Rechecking the consistancy
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

## Identifying Target(y) and Feature(X)

In [198]:
# specifying the target
y= df['Item_Outlet_Sales']
# specifying X features
X = df.drop(columns ='Item_Outlet_Sales')
X.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  
0  Supermarket Type1  
1  Supermarket Type2  
2  Supermarket Type1  
3      Grocery Store  
4  Supermarket Type1

In [199]:
X['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [200]:
#perfoming the split
X_train, X_test,y_train, y_test = train_test_split(X,y,random_state=42)


## Performing PreProcessing

### Pipeline for preprocessing numeric columns

In [201]:
# identify Numeric columns
X_train_NumCol  = X_train.select_dtypes('number').columns
X_train_NumCol

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')

In [202]:
#checking missing values
X_train[X_train_NumCol].describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
count      5285.00          6392.00   6392.00                    6392.00
mean         12.90             0.07    141.98                    1997.86
std           4.64             0.05     62.63                       8.39
min           4.56             0.00     31.29                    1985.00
25%           8.90             0.03     94.15                    1987.00
50%          12.65             0.05    144.11                    1999.00
75%          17.00             0.09    186.90                    2004.00
max          21.35             0.33    266.89                    2009.00

Item_weight has a missing values we are going to impute with a strategy = mean

In [203]:
#instantiating the imputer
num_imputer =SimpleImputer(strategy='mean')
num_scaler = StandardScaler()


In [204]:
#Create pipeline
num_pipe =make_pipeline(num_imputer,num_scaler)
num_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler())])

In [205]:
#Creating tuples for num_col
num_tuple =('numeric',num_pipe,X_train_NumCol)
num_tuple

('numeric',
 Pipeline(steps=[('simpleimputer', SimpleImputer()),
                 ('standardscaler', StandardScaler())]),
 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
        'Outlet_Establishment_Year'],
       dtype='object'))

### Pipeline for Preprocessing Nominal Feature

In [206]:
# Identifying Nominal Features
cat_col =X_train.select_dtypes('object').columns
cat_col

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [207]:
# Checking missing values
X_train[cat_col].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6392 entries, 4776 to 7270
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Item_Identifier       6392 non-null   object
 1   Item_Fat_Content      6392 non-null   object
 2   Item_Type             6392 non-null   object
 3   Outlet_Identifier     6392 non-null   object
 4   Outlet_Size           4580 non-null   object
 5   Outlet_Location_Type  6392 non-null   object
 6   Outlet_Type           6392 non-null   object
dtypes: object(7)
memory usage: 399.5+ KB


There are no missing values in categorical features so no need to impute

In [208]:
# instantiate the one-hot-encoder
cat_one_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')


In [209]:
# Creating pipeline
cat_pipe  =make_pipeline(cat_one_encoder)
cat_pipe


Pipeline(steps=[('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [210]:
# creating tuple
cat_tuple  =('categorical',cat_pipe,cat_col)
cat_tuple

('categorical',
 Pipeline(steps=[('onehotencoder',
                  OneHotEncoder(handle_unknown='ignore', sparse_output=False))]),
 Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
        'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
       dtype='object'))

### Instantiating  Column Transformer

In [211]:
# instantiate ColumnTransformer

col_transformer = ColumnTransformer([num_tuple, cat_tuple],verbose_feature_names_out=False)

# fit the data to Transformer
col_transformer.fit(X_train)


ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [212]:
X_train_procced = col_transformer.transform(X_train)
X_train_procced.round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776         0.82            -0.71      1.83                       1.33   
7510         0.56            -1.29      0.60                       1.33   
5828        -0.13             1.81      0.24                       0.14   
5327        -1.17            -1.00     -0.95                       0.73   
4810         1.53            -0.97     -0.34                       0.49   
...           ...              ...       ...                        ...   
5734        -0.83             4.31     -0.04                       0.02   
5191         0.64             1.01     -1.06                       1.09   
5390         1.11            -0.92      1.52                       0.49   
860          1.77            -0.23     -0.38                       1.09   
7270         0.82            -0.96     -0.74                      -0.10   

      Item_Identifier_DRA12  Item_Identifier_DRA24  Item_Identifier_DRA59  \
4776                    0.0                    0.0                    0.0   
7510                    0.0                    0.0                    0.0   
5828                    0.0                    0.0                    0.0   
5327                    0.0                    0.0                    0.0   
4810                    0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
5734                    0.0                    0.0                    0.0   
5191                    0.0                    0.0                    0.0   
5390                    0.0                    0.0                    0.0   
860                     0.0                    0.0                    0.0   
7270                    0.0                    0.0                    0.0   

      Item_Identifier_DRB01  Item_Identifier_DRB13  Item_Identifier_DRB24  \
4776                    0.0                    0.0                    0.0   
7510                    0.0                    0.0                    0.0   
5828                    0.0                    0.0                    0.0   
5327                    0.0                    0.0                    0.0   
4810                    0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
5734                    0.0                    0.0                    0.0   
5191                    0.0                    0.0                    0.0   
5390                    0.0                    0.0                    0.0   
860                     0.0                    0.0                    0.0   
7270                    0.0                    0.0                    0.0   

      ...  Outlet_Size_Medium  Outlet_Size_Small  Outlet_Size_nan  \
4776  ...                 1.0                0.0              0.0   
7510  ...                 1.0                0.0              0.0   
5828  ...                 1.0                0.0              0.0   
5327  ...                 0.0                1.0              0.0   
4810  ...                 0.0                0.0              1.0   
...   ...                 ...                ...              ...   
5734  ...                 0.0                0.0              1.0   
5191  ...                 0.0                0.0              1.0   
5390  ...                 0.0                0.0              1.0   
860   ...                 0.0                0.0              1.0   
7270  ...                 0.0                1.0              0.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
4776                          0.0                          0.0   
7510                          0.0                          0.0   
5828                          1.0                          0.0   
5327                          0.0                          1.0   
4810                          0.0                          1.0   
...                           

In [213]:
# Checking if null still exists
X_train_procced.isna().sum().sum()

0

In [214]:
# transforming X_test
X_test_processed =col_transformer.transform(X_test)
X_test_processed.round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
7503         0.33            -0.78     -1.00                      -1.29   
2957        -1.18             0.10     -1.59                      -0.10   
7031         0.38            -0.48     -1.60                       0.14   
1084         0.00            -0.42      0.51                      -1.53   
856         -0.64            -1.05      0.89                       0.73   
...           ...              ...       ...                        ...   
4753         1.11            -1.13      0.47                      -1.29   
4836        -0.64            -1.29      0.02                       1.09   
8064        -1.14             1.22      1.09                       0.49   
4418        -1.50            -0.78     -0.37                       0.14   
1784         0.52            -0.78      0.11                       1.09   

      Item_Identifier_DRA12  Item_Identifier_DRA24  Item_Identifier_DRA59  \
7503                    0.0                    0.0                    0.0   
2957                    0.0                    0.0                    0.0   
7031                    0.0                    0.0                    0.0   
1084                    0.0                    0.0                    0.0   
856                     0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
4753                    0.0                    0.0                    0.0   
4836                    0.0                    0.0                    0.0   
8064                    0.0                    0.0                    0.0   
4418                    0.0                    0.0                    0.0   
1784                    0.0                    0.0                    0.0   

      Item_Identifier_DRB01  Item_Identifier_DRB13  Item_Identifier_DRB24  \
7503                    0.0                    0.0                    0.0   
2957                    0.0                    0.0                    0.0   
7031                    0.0                    0.0                    0.0   
1084                    0.0                    0.0                    0.0   
856                     0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
4753                    0.0                    0.0                    0.0   
4836                    0.0                    0.0                    0.0   
8064                    0.0                    0.0                    0.0   
4418                    0.0                    0.0                    0.0   
1784                    0.0                    0.0                    0.0   

      ...  Outlet_Size_Medium  Outlet_Size_Small  Outlet_Size_nan  \
7503  ...                 0.0                0.0              0.0   
2957  ...                 0.0                1.0              0.0   
7031  ...                 1.0                0.0              0.0   
1084  ...                 1.0                0.0              0.0   
856   ...                 0.0                1.0              0.0   
...   ...                 ...                ...              ...   
4753  ...                 0.0                0.0              0.0   
4836  ...                 0.0                0.0              1.0   
8064  ...                 0.0                0.0              1.0   
4418  ...                 1.0                0.0              0.0   
1784  ...                 0.0                0.0              1.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
7503                          0.0                          0.0   
2957                          1.0                          0.0   
7031                          1.0                          0.0   
1084                          0.0                          0.0   
856                           0.0                          1.0   
...                           

there are no more null values , data has been scaled successfully